README.md

# rytstat - R interface for work with YouTube APIs <a href='https://selesnow.github.io/rytstat/'><img src='man/figures/logo.png' align="right" height="138.5" /></a>

<!-- badges: start -->
[![](https://cranlogs.r-pkg.org/badges/grand-total/rytstat)](https://cran.r-project.org/package=rytstat)
[![](https://cranlogs.r-pkg.org/badges/rytstat?color=lightgrey)](https://cran.r-project.org/package=rytstat)
[![](https://cranlogs.r-pkg.org/badges/last-week/rytstat?color=lightgrey)](https://cran.r-project.org/package=rytstat)
[![CRAN status](https://www.r-pkg.org/badges/version/rytstat)](https://CRAN.R-project.org/package=rytstat)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/selesnow/rytstat/workflows/R-CMD-check/badge.svg)](https://github.com/selesnow/rytstat/actions)
<!-- badges: end -->

`rytstat` package is an R interface for working with the following YouTube APIs:

* [YouTube Reporting API](https://developers.google.com/youtube/reporting/v1/reports)
* [YouTube Analytics API](https://developers.google.com/youtube/analytics/data_model)
* [YouTube Data API](https://developers.google.com/youtube/v3/getting-started)

That is, the `rytstat` allows you to request any data available in the YouTube Creator Studio for further analysis and visualization using the R language.

## Privacy Policy

The `rytstat` package for authorization uses the [gargle](https://gargle.r-lib.org/) package, the credentials obtained during authorization are stored exclusively on your local PC, you can find out the folder into which the credentials are cached using the `ryt_auth_cache_path()` function.

For loading data from your YouTube channel `rytstat` needs next scopes:

* View monetary and non-monetary YouTube Analytics reports for your YouTube content
* View your YouTube account
* View and manage your assets and associated content on YouTube
* View YouTube Analytics reports for your YouTube content
* Manage your YouTube account

For more details see [Official YouTube API documentation](https://developers.google.com/youtube/reporting/guides/authorization#identify-access-scopes).

The package does not transfer your credentials or data obtained from your advertising accounts to third parties, however, the responsibility for information leakage remains on the side of the package user. The author does not bear any responsibility for their safety, be careful when transferring cached credentials to third parties.

For more details, I recommend that you read the following articles from the official documentation of the gargle package:

* [Stewarding the cache of user tokens](https://www.tidyverse.org/blog/2021/07/gargle-1-2-0/)
* [Auth when using R in the browser](https://cran.r-project.org/package=gargle/vignettes/auth-from-web.html)
* [How gargle gets tokens](https://cran.r-project.org/package=gargle/vignettes/how-gargle-gets-tokens.html)

### Authorization process

You run `gads_auth('me@gmail.com')` and start [OAuth Dance](https://medium.com/typeforms-engineering-blog/the-beginners-guide-to-oauth-dancing-4b8f3666de10) in the browser:

![Typical OAuth dance in the browser, when initiated from within R](https://raw.githubusercontent.com/selesnow/rytstat/master/man/figures/auth_process.png)

Upon success, you see this message in the browser:

`Authentication complete. Please close this page and return to R.`

And you credentials cached locally on your PC in the form of RDS files.

### Key points
* By default, gargle caches user tokens centrally, at the user level, and their keys or labels also convey which Google identity is associated with each token.
* Token storage relies on serialized R objects. That is, tokens are stored locally on your PC in the form of RDS files.

### Use own OAuth client
You can use own OAuth app:

```r
app <- httr::oauth_app(appname = "app name", key = "app id", secret = "app secret")
ryt_auth_configure(app = app)

# or from json file 
ryt_auth_configure(path = 'D:/ga_auth/app.json')

# run authorization
ryt_auth('me@gmail.com')
```

## Install

You can install `rytstat` from [CRAN](https://cran.r-project.org/package=rytstat) or [GitHub](https://github.com/selesnow/rytstat):

``` r
install.packages("rytstat")
```

или GitHub:

```r
devtools::install_github('selesnow/rytstat')
```

## Examples
### Auth

```r
library(rytstat)
library(httr)

# auth app
app <- oauth_app(
    appname = 'my app',
    key = 'app id', 
    secret = 'app secret')

ryt_auth_configure(app = app)

ryt_auth(email = 'me@gmail.com')


```

### Using YouTube Data API

``` r
# load channel data
channel <- ryt_get_channels()

# load videos
videos <- ryt_get_video_list()
video_details <- ryt_get_video_details(video_id = videos$id_video_id)

# load playlists
pl <- ryt_get_playlists()
pl_items <- ryt_get_playlist_items(
     playlist_id = pl$id[1],
     part = c('contentDetails', 'snippet'),
     fields = 'items(id,snippet/channelId,snippet/title,contentDetails/videoId)'
)

# search channels, playlists or videos
search_res_videos <- ryt_search(
    type            = 'video',
    q               = 'r language tutorial',
    published_after = '2022-03-01T00:00:00Z',
    published_before = '2022-06-01T00:00:00Z',
    max_results     = 10 
)

search_res_playlists <- ryt_search(
    type             = 'playlist',
    q                = 'r language tutorial',
    published_after  = '2022-03-01T00:00:00Z',
    published_before = '2022-06-01T00:00:00Z',
    max_results      = 50
)

search_res_channels <- ryt_search(
    type             = 'channel',
    q                = 'r language tutorial',
    published_after  = '2022-03-01T00:00:00Z',
    published_before = '2022-06-01T00:00:00Z',
    max_results      = 50
)
```

### Using YouTube Analytics API


``` r
library(rytstat)

# get list of videos
videos <- ryt_get_video_list()

# get statistics by day and videos
# you can specify no more than 500 videos at a time
video_stat <- ryt_get_analytics(
  start_date = '2021-01-01', 
  end_date = '2021-09-01',
  dimensions = c('day', 'video'),
  metrics = c('views', 
              'likes', 
              'dislikes', 
              'comments', 
              'shares'),
  filters = str_glue('video=={str_c(head(videos$id_video_id, 500), collapse=",")}')
)

```

### Using YouTube Reporting API

```r
# auth
ryt_auth('me@gmail.com')

# get reporting data
## create job
ryt_create_job('channel_basic_a2')

## get job list
jobs2 <- ryt_get_job_list()

## get job report list
reports <- ryt_get_report_list(
  job_id = jobs$id[1],
  created_after = '2021-10-20T15:01:23.045678Z'
)

# get report data
data <- ryt_get_report(
  download_url = reports$downloadUrl[1]
)

# delete job
ryt_delete_job(jobs$id[1])
```

## Author
Alexey Seleznev, Head of analytics dept. at [Netpeak](https://netpeak.net/en/us/)
<Br>Telegram Channel: [R4marketing](https://t.me/R4marketing)
<Br>YouTube Channel: [R4marketing](https://www.youtube.com/R4marketing/?sub_confirmation=1)
<Br>email: selesnow@gmail.com
<Br>facebook: [facebook.com/selesnow](https://www.facebook.com/selesnow)
<Br>blog: [alexeyseleznev.wordpress.com](https://alexeyseleznev.wordpress.com/)



NEWS.md
# rytstat (development version)
* В функцию `ryt_get_videos()` добавлен аргумент `fields`, позволяющий указать нужные поля.
* В функцию `ryt_get_analytics()` добавлены следующие аргументы:
    * `sort` - Сортировка данных в отчёте
    * `max_results` - Ограничение на количество строк в отчёте
    * `start_index` - Начальный индекс строки, с которой необходимо получить данные.

Новые аргументы функции `ryt_get_analytics()` позволяют запрашивать топ видео за указанный период.
```r
# top video
  top_10_by_view <- ryt_get_analytics(
    '2022-06-01', '2022-06-30', 
    metrics = 'views',
    dimensions = 'video',
    sort = '-views', 
    max_results = 10
  )
  
```
* Исправлена ошибка в функции `ryt_get_report_list()`, изза которой не работали аргументы `created_after`, `start_time_at_or_after`, `start_time_before`.
* В случае если запрос `ryt_get_report_list()` не вернул никаких результатов, работа функции будет остановлена, а сама функция вернёт `NULL`. Ранее в таком случае функция возвращала таблицу из одного безымянного столбца и пустой строки.

# rytstat 0.3.0

## Новая функция
* В пакет добавлена функция `ryt_search()`, предназначенная для поиска видео, плей листов и каналов, не обязательно вашего авторства.

## Основное
* В функциях `ryt_get_video_details()`, `ryt_get_playlists()`, `ryt_get_channel_activities()`, `ryt_get_channels()` и `ryt_get_playlist_items()` аргумент `fields` переименован в `part`, и добавлен аргумент `fields`, который позволяет запрашивать отдельные поля частей запроса. Примеры работы с новыми аргументами добавлены в справку функции.

## Документация
* Расширена документации функций `ryt_get_video_details()`, `ryt_get_playlists()`, `ryt_get_channel_activities()`, `ryt_get_channels()` и `ryt_get_playlist_items()` в неё добавлены примеры с использованием аргумента `fields`, и добавлен перечень частей (parts) и полей (fields).

## Другое
* Вшитый в `rytstat` OAuth клиент был подтверждён Google, теперь все пользователи пакета могут использовать его для авторизациию

# rytstat 0.2.1

## Документация
* Исправлены ссылки в README и DESCRIPTION

# rytstat 0.2.0

## Новые функции
* Доработал функцию `ryt_get_playlists()` таким образом, что бы она разворачивала все вложенные столбцы.
* Добавлена функция `ryt_get_playlist_items()`, которая позволяет запросить элементы плейлиста YouTube.
* Добавлена функция `ryt_get_channels()`, которая позволяет получить метаданные по каналу.
* Добавдена функция `ryt_get_channel_activities()`, которая позволяет получить список активностей связанных с вашим каналом.
* Добавлена функция `ryt_get_comments()`, которая позволяет запросить список комментариев к видео или каналу.

## Исправления ошибок
* Исправлено поведение функции `ryt_get_analytics()`, ранее нельзя было изменить параметр `dimensions`, он всегда при запросе был равен 'day'.

## Названия функций
* Следующие функции были переименованы с целью сокращения их названий:
    * `ryt_reports_create_job()`      -> `ryt_create_job()`
    * `ryt_reports_get_job_list()`    -> `ryt_get_job_list()`
    * `ryt_reports_get_report_list()` -> `ryt_get_report_list()`
    * `ryt_reports_delete_job()`      -> `ryt_delete_job()`
    * `ryt_get_video_list()`          -> `ryt_get_videos()`
    
## Документация
* В документацию всех функций добавлены примеры кода. 
* Создан официальный сайт пакета `rytstat` - https://selesnow.github.io/rytstat/docs

## Прочее
* Начиная с версии 0.2.0 пакет требует разрешение https://www.googleapis.com/auth/youtube.force-ssl, которое позволяет rytstat просматривать комментарии.

# rytstat 0.1.0

* Добавлен `NEWS.md` для фиксации изменений в пакете.
* Пакет опубликован на CRAN.


R/ryt_search.R
#' Search channels, playlists or videos on YouTube
#' @description Returns a collection of search results that match the query parameters specified in the API request. By default, a search result set identifies matching video, channel, and playlist resources, but you can also configure queries to only retrieve a specific type of resource.
#' @param part The part parameter specifies a comma-separated list of one or more search resource properties that the API response will include. Set the parameter value to snippet.
#' @param ... Filters and Optional parameters for resources search. See [list of all allowed params](https://developers.google.com/youtube/v3/docs/search/list#parameters). You can set params names in came or snake case.
#'
#' @return resources list
#' @export
#'
#' @examples
#' \dontrun{
#' # search videos by query
#' search_res_videos <- ryt_search(
#'   type            = 'video',
#'   q               = 'r language tutorial',
#'   published_after = '2022-03-01T00:00:00Z',
#'   published_before = '2022-06-01T00:00:00Z',
#'   max_results     = 10
#' )
#'
#' # search playlists by query
#' search_res_playlists <- ryt_search(
#'   type             = 'playlist',
#'   q                = 'r language tutorial',
#'   published_after  = '2022-03-01T00:00:00Z',
#'   published_before = '2022-06-01T00:00:00Z',
#'   max_results      = 50
#' )
#'
#' # search channel by query
#' search_res_channels <- ryt_search(
#' type             = 'channel',
#' q                = 'r language tutorial',
#' published_after  = '2022-03-01T00:00:00Z',
#'   published_before = '2022-06-01T00:00:00Z',
#'   max_results      = 50
#' )
#'
#' # Search in own videos
#' search_own_dplyr_videos <- ryt_search(
#'   type             = 'video',
#'   for_mine         = TRUE,
#'   q                = 'dplyr'
#' )
#'
#' # Search channels and get title and id
#' search_chn <-  ryt_search(
#'    type   = 'channel',
#'    q      = 'R4marketing',
#'    fields = 'items(snippet(title,channelId))'
#' )
#'
#' # Search videos in the channel by query and channel id
#' search_channel_dplyr_videos <- ryt_search(
#'   type       = 'video',
#'   q          = 'dplyr',
#'   channel_id = "UCyHC6R3mCCP8bhD9tPbjnzQ"
#' )
#' }
ryt_search <- function(
    part = "snippet",
    ...
  ) {

  cli_alert_info('Compose params')
  part <- paste0(part, collapse = ",")
  q_params <- c(as.list(environment()), list(...))
  names(q_params) <- names(q_params) %>% to_lower_camel_case()
  result <- list()

  cli_alert_info('Send query')

  while (!is.null(q_params$pageToken)|!exists('resp', inherits = FALSE)) {

    out <- request_build(
      method   = "GET",
      params   = q_params,
      token    = ryt_token(),
      path     = 'youtube/v3/search',
      base_url = 'https://www.googleapis.com/'
    )

    # send request
    ans <- request_retry(
      out,
      encode = 'json'
    )

    resp <- response_process(ans)

    result <- append(result, list(resp$items))

    q_params$pageToken <- resp$nextPageToken

  }

  cli_alert_info('Parse result')
  result <- tibble(items = result) %>%
            unnest_longer(.data$items) %>%
            unnest_wider(.data$items)

  nested_fields <- select(result, where(is.list)) %>% names()
  nested_fields <- nested_fields[!nested_fields %in% c("tags", "topicDetails")]

  while ( length(nested_fields) > 0 ) {

    for ( col in nested_fields ) {

      if (col == "tags") next

      result_t <- try(unnest_wider(result, any_of(col)), silent = T)

      if ( 'try-error' %in% class(result_t) ) {
        result <- unnest_wider(result, any_of(col), names_sep = '_')
      } else {
        result <- result_t
      }

    }

    nested_fields <- select(result, where(is.list)) %>%
                     names()

    nested_fields <- nested_fields[!nested_fields %in% c("tags", "topicDetails")]

  }

  cli_alert_success(str_glue('Success, loading {nrow(result)} rows.'))
  return(result)

}

